### PR TITLE
Added a var for Viewer.

### LINKED ISSUE #1493: Expose viewer data to amp-analytics
amp-analytics currently has support for the DOCUMENT_REFERRER substitution. This request covers exposing viewer data like the name to amp-analytics so that a publisher could know the viewing context for the page. For instance, consider a user who does the following:

(a) Sees a link on Google clicks on it, opening an article on pub.com.
(b) While reading the article in (a), clicks on another link to a page but stays within the Google viewing context.

At the end of (a) we'd want to have DOCUMENT_REFERRER = google.com and VIEWER = Google. At the end of (b) we'd want to have DOCUMENT_REFERRER = pub.com and VIEWER = Google.

### PR DESCRIPTION
**Pull request by avimehta**
This var is returns the hostname found in the 'origin' param of the
viewer or the referrer.

There are a couple of open issues in this PR. Comments and suggestions welcome.

1. The code currently reads the origin param from the iframe and parses the domain out of it. Should it return the value without any processing instead? That way, apps don't have to pass in a domain or a URL and can instead just pass in a string.
2. The code returns the referrer information if the origin is not found. I am not sure if that should be done or if it makes sense.

Fixes #1493

### CODE REVIEW

File: src/url-replacements.js [hunk 1]
```diff
@@ -29,6 +29,9 @@ import {userNotificationManagerFor} from './user-notification';
 /** @private {string} */
 const TAG_ = 'UrlReplacements';
 
+/** @private {string} */
+const VIEWPORT_ORIGIN_PARAM_ = 'origin';
```
Comment (dvoytenko) [inline src/url-replacements.js]: This should probably be "VIEWER", not a "VIEWPORT"

File: src/url-replacements.js [hunk 2]
```diff
@@ -239,6 +242,14 @@ class UrlReplacements {
       return this.getTimingData_('navigationStart',
           'domContentLoadedEventStart');
     });
+
+    // Returns an identifier for the viewer.
+    this.set_('VIEWER', () => {
+      const viewerUrl = viewerFor(this.win_).getParam(VIEWPORT_ORIGIN_PARAM_) ||
```
Comment (dvoytenko) [inline src/url-replacements.js]: This implementation should go into `viewer-impl.js`, there's already `messagingOrigin_` there that can be used. Also, `getReferrerUrl` returns a promise, so this implementation is somewhat incorrect. Also, `getReferrerUrl` might be a good idea here, but could you explain why?

File: test/functional/test-url-replacements.js [hunk 3]
```diff
@@ -19,17 +19,30 @@ import {createIframePromise} from '../../testing/iframe';
 import {urlReplacementsFor} from '../../src/url-replacements';
 import {markElementScheduledForTesting} from '../../src/custom-element';
 import {installCidService} from '../../src/service/cid-impl';
+import {installViewerService} from '../../src/service/viewer-impl';
 import {setCookie} from '../../src/cookies';
-
+import * as sinon from 'sinon';
```
Comment (jridgewell) [inline test/functional/test-url-replacements.js]: We shouldn't need to import `sinon`, it will be available in the global scope.

Comment (dvoytenko) [inline test/functional/test-url-replacements.js]: There was some weird problem with it and fake timers. Not sure if that was addressed on sinon side. @cramforce do you remember?

File: src/url-replacements.js (hunk 2 above)
Comment (avimehta) [inline src/url-replacements.js]: messagingOrigin_ is only set once setMessageDeliverer is called. How do I ensure that it is called before i look for id? 

I am not sure if getReferrerUrl makes sense (that's why the request for comments in original PR). I added it here for the case when origin is not specified. I see that there is a todo in codebase to make origin required. If that is the case, I could remove the referrer part.

/cc @rudygalfi

File: extensions/amp-analytics/analytics-vars.md [hunk 4]
```diff
@@ -74,6 +74,12 @@ Provides the canonical URL of the current document.
 
 Example value: `http%3A%2F%2Fexample.com%3A8000%2Fanalytics.html`
 
+### documentCharset
```
Comment (rudygalfi) [inline extensions/amp-analytics/analytics-vars.md]: Just wondering if this is in here because we forgot to document originally?

File: extensions/amp-analytics/analytics-vars.md [hunk 5]
```diff
@@ -86,6 +92,11 @@ Provides the title of the current document.
 
 Example value: `The New York Times - Breaking News, World News...`
 
+### viewer
+Provides an identifier for the host of the AMP document. Empty string if the document is loaded directly in a browser.
```
Comment (rudygalfi) [inline extensions/amp-analytics/analytics-vars.md]: Might want to reconsider the word "host". A bit overloaded. Will think of alternatives.

File: extensions/amp-analytics/analytics-vars.md (hunk 4 above)
Comment (rudygalfi) [inline extensions/amp-analytics/analytics-vars.md]: Oh, I see. It just got moved around.

File: spec/amp-var-substitutions.md [hunk 6]
```diff
@@ -95,6 +104,14 @@ For instance:
 ```
 may make a request to something like `https://foo.com/pixel?title=Breaking%20News`.
 
+### VIEWER
+
+Provides an identifier for the host of the AMP document. Empty string is provided when the document is loaded directly in the browser.
```
Comment (rudygalfi) [inline spec/amp-var-substitutions.md]: same feedback re: "host"

File: src/url-replacements.js (hunk 2 above)
Comment (dvoytenko) [inline src/url-replacements.js]: `messagingOrigin_` is easy - you simply need to wrap it in a promise. E.g. you will have an open promise that will be resolved in the same place where `messagingOrigin_` is set.

RE: viewer vs referrer - please touch base with @jridgewell - he dealt with this problem. I believe what he did was to translate a set of referrers to viewers. But I could be wrong.

File: test/functional/test-url-replacements.js (hunk 3 above)
Comment (dvoytenko) [inline test/functional/test-url-replacements.js]: Confirmed. We have to do this for now.

File: src/url-replacements.js (hunk 2 above)
Comment (jridgewell) [inline src/url-replacements.js]: I had to translate from a (small) set of UserAgents into a referrer since some services do not provide a referrer.

But we need to keep the distinction between viewer referrers (maybe "viewer origin"?) vs referrer. When you open an AMP doc (let's call it "A") from the SERP viewer, both the viewer referrer and referrer should be "google.com". If we were to follow some link in that doc to another AMP doc (called "B", and the viewer could load it), the viewer referrer should still be "google.com" but the referrer should be doc A.

Comment (avimehta) [inline src/url-replacements.js]: ok.. so remove the referrer part from calculation of viewer origin? In that case, how do we ensure that origin is always specified?

Comment (avimehta) [inline src/url-replacements.js]: Also, do we expect apps to also specify a URL in the origin parameter?

Comment (dvoytenko) [inline src/url-replacements.js]: Unfortunately, this is not a fully solvable problem. When the page is iframe'd and comms set up, we have the right value via `messagingOrigin_`. I think we need to start here and build on this as we encounter case by case.

Comment (dvoytenko) [inline src/url-replacements.js]: Checking for `origin` in the URL is not a great idea because it's very easily spoofable. That's why we have `messagingOrigin_` which is set via handshake. It all depends on the sensitivity of the problem. For instance, we have both: unconfirmed and confirmed referrers. The reason is because spoofing a referrer is not always a risk. How much risk here is to spoof viewer origin for your needs?

Comment (avimehta) [inline src/url-replacements.js]: hmm. That complicates things if apps want to host a amp doc. How will apps do this handshake? 

For analytics, client id is also dependent on the origin being trusted. I'll start with just using the value of `messagingOrigin_`. Is that always set in case of google.com?

Comment (dvoytenko) [inline src/url-replacements.js]: There's no real point of doing handshake in a native app - it's very easily spoofed and there's no way to confirm.

I'd expect `messagingOrigin_` to be always set in a iframed scenario. Native apps - we don't yet have a way to always provide it. Feel free to file a bug, e.g. "Propagate viewer info from native app".

Let's start with exposing messagingOrigin_ as a promise. Let's call it `getViewerOrigin()`. I'll provide other details in the PR review itself.

File: src/url-replacements.js (hunk 1 above)
Comment (avimehta) [inline src/url-replacements.js]: done.

File: src/url-replacements.js (hunk 2 above)
Comment (avimehta) [inline src/url-replacements.js]: done as discussed. I am not very confident about promises ;). Can you take extra care going through the changes there?

File: extensions/amp-analytics/analytics-vars.md (hunk 4 above)
Comment (avimehta) [inline extensions/amp-analytics/analytics-vars.md]: yes, the documentCharset was in the wrong category. I moved it to the correct one. I can send out a separate PR if that is preferred.

File: extensions/amp-analytics/analytics-vars.md (hunk 5 above)
Comment (avimehta) [inline extensions/amp-analytics/analytics-vars.md]: does this look better?

**Comment by avimehta** [comment]
ptal.

File: spec/amp-var-substitutions.md [hunk 7]
```diff
@@ -95,6 +104,14 @@ For instance:
 ```
 may make a request to something like `https://foo.com/pixel?title=Breaking%20News`.
 
+### VIEWER
+
+Provides an identifier for the viewer that contains the AMP document. Empty string is provided when the document is loaded directly in the browser or if the id is not found.
+For instance:
+```html
+<amp-pixel src="https://foo.com/pixel?title=VIEWER"></amp-pixel>
```
Comment (dvoytenko) [inline spec/amp-var-substitutions.md]: nit: `title=VIEWER` -> `viewer=VIEWER`

File: src/service/viewer-impl.js [hunk 8]
```diff
@@ -154,9 +154,6 @@ export class Viewer {
     /** @private {?function(string, *, boolean):(Promise<*>|undefined)} */
     this.messageDeliverer_ = null;
 
-    /** @private {?string} */
-    this.messagingOrigin_ = null;
```
Comment (dvoytenko) [inline src/service/viewer-impl.js]: We need to keep `this.messaingOrigin_` var intact - it will be needed for other purposes. We'll simply create a new set of tools for viewer origin.

File: src/service/viewer-impl.js [hunk 9]
```diff
@@ -257,6 +254,13 @@ export class Viewer {
     /** @const @private {!Promise<boolean>} */
     this.isTrustedViewer_ = trustedViewerPromise;
 
+    const messagingOriginPromise = new Promise(resolve => {
+      /** @const @private {!function(string)} */
+      this.messagingOriginResolver_ = resolve;
+    });
+    /** @const @private {!Promise<string>} */
+    this.messagingOrigin_ = messagingOriginPromise;
```
Comment (dvoytenko) [inline src/service/viewer-impl.js]: Let's make this code look something like this:

```
this.viewerOrigin_ = new Promise(resolve => {
  if (!this.isEmbedded_) {
    // Viewer is only determined for iframed documents at this time.
    resolve('');
  } else if (this.win.location.ancestorOrigins &&
      this.win.location.ancestorOrigins.length > 0) {
    resolve(this.win.location.ancestorOrigins[0]);
  } else {
    // Wait until other signals arrive such as messaging origin.
    /** @private @const {!function(string)|undefined} */
    this.viewerOriginResolver_ = resolve;
  }
});
```

File: src/service/viewer-impl.js [hunk 10]
```diff
@@ -495,6 +499,14 @@ export class Viewer {
   }
 
   /**
+   * Returns the promise that resolves to URL representing the origin of the viewer.
+   * @return {Promise<string>}
+   */
+  getViewerOrigin() {
+    return this.messagingOrigin_;
```
Comment (dvoytenko) [inline src/service/viewer-impl.js]: Thus, this will become `return this.viewerOrigin_`.

File: src/service/viewer-impl.js [hunk 11]
```diff
@@ -495,6 +499,14 @@ export class Viewer {
   }
 
   /**
+   * Returns the promise that resolves to URL representing the origin of the viewer.
```
Comment (dvoytenko) [inline src/service/viewer-impl.js]: Document possible return values/situations. E.g. iframed vs non-iframed case.

File: src/service/viewer-impl.js [hunk 12]
```diff
@@ -679,11 +691,14 @@ export class Viewer {
     log.fine(TAG_, 'message channel established with origin: ', origin);
     this.messageDeliverer_ = deliverer;
     // TODO(dvoytenko, #1764): Make `origin` required when viewers catch up.
-    this.messagingOrigin_ = origin;
     if (this.trustedViewerResolver_) {
       this.trustedViewerResolver_(
           origin ? this.isTrustedViewerOrigin_(origin) : false);
     }
+    if (this.messagingOriginResolver_) {
```
Comment (dvoytenko) [inline src/service/viewer-impl.js]: So, let's revert `this.messagingOrigin_ = origin;` and instead add `this.viewerOriginResolver_(origin || '')`.

File: src/service/viewer-impl.js (hunk 9 above)
Comment (dvoytenko) [inline src/service/viewer-impl.js]: We also need tests for this in `test-viewer.js`. It should be very similar to test-set for `getReferrerUrl`.

File: src/url-replacements.js [hunk 13]
```diff
@@ -239,6 +239,14 @@ class UrlReplacements {
       return this.getTimingData_('navigationStart',
           'domContentLoadedEventStart');
     });
+
+    // Returns an identifier for the viewer.
+    this.set_('VIEWER', () => {
+      return viewerFor(this.win_).getViewerOrigin().then(origin => {
+        const url = parseUrl(origin);
+        return (url && url.hostname) || '';
```
Comment (dvoytenko) [inline src/url-replacements.js]: One thing to consider - we can return origin for `VIEWER` var, e.g. not `www.google.com` but `https://www.google.com`. LMK your thoughts.

File: src/url-replacements.js [hunk 14]
```diff
@@ -239,6 +239,14 @@ class UrlReplacements {
       return this.getTimingData_('navigationStart',
           'domContentLoadedEventStart');
     });
+
+    // Returns an identifier for the viewer.
+    this.set_('VIEWER', () => {
+      return viewerFor(this.win_).getViewerOrigin().then(origin => {
+        const url = parseUrl(origin);
```
Comment (dvoytenko) [inline src/url-replacements.js]: Origin can be empty, so let's handle here. Otherwise it will resolve relative URL against the empty string.

File: test/functional/test-url-replacements.js [hunk 15]
```diff
@@ -272,6 +291,16 @@ describe('UrlReplacements', () => {
     });
   });
 
+  it('should replace VIEWER with origin', () => {
+    return getReplacements().then(replacements => {
+      sandbox.stub(viewerService, 'getViewerOrigin').returns(
```
Comment (dvoytenko) [inline test/functional/test-url-replacements.js]: Add a test where getViewerOrigin resolves to an empty string.

File: extensions/amp-analytics/0.1/vendors.js [hunk 16]
```diff
@@ -54,6 +54,7 @@ export const ANALYTICS_CONFIG = {
       'timestamp': 'TIMESTAMP',
       'timezone': 'TIMEZONE',
       'title': 'TITLE',
+      'viewer': 'VIEWER',
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/vendors.js]: Do we have any sort of tests for these bindings?

File: src/url-replacements.js (hunk 13 above)
Comment (avimehta) [inline src/url-replacements.js]: @rudygalfi  what do you think? i am fine with that. that also works if in future, viewerOrigin stops being URL (for example for apps.)

File: extensions/amp-analytics/0.1/vendors.js (hunk 16 above)
Comment (avimehta) [inline extensions/amp-analytics/0.1/vendors.js]: No we do not. repurposed https://github.com/ampproject/amphtml/issues/1779

File: test/functional/test-url-replacements.js (hunk 15 above)
Comment (avimehta) [inline test/functional/test-url-replacements.js]: done.

File: spec/amp-var-substitutions.md (hunk 7 above)
Comment (avimehta) [inline spec/amp-var-substitutions.md]: done.

File: src/service/viewer-impl.js (hunk 8 above)
Comment (avimehta) [inline src/service/viewer-impl.js]: done.

File: src/service/viewer-impl.js (hunk 9 above)
Comment (avimehta) [inline src/service/viewer-impl.js]: Can you suggest more tests. Added a few.

File: src/service/viewer-impl.js (hunk 10 above)
Comment (avimehta) [inline src/service/viewer-impl.js]: ack.

File: src/service/viewer-impl.js (hunk 11 above)
Comment (avimehta) [inline src/service/viewer-impl.js]: done.

File: src/service/viewer-impl.js (hunk 12 above)
Comment (avimehta) [inline src/service/viewer-impl.js]: done.

File: src/url-replacements.js (hunk 14 above)
Comment (avimehta) [inline src/url-replacements.js]: done.

**Comment by avimehta** [comment]
Not sure why @dvoytenko 's commit got involved here. Will try to fix it but this happened in someone else's PR as well today.

File: src/service/viewer-impl.js [hunk 17]
```diff
@@ -157,6 +164,9 @@ export class Viewer {
     /** @private {?string} */
     this.messagingOrigin_ = null;
 
+    /** @private {?string} */
+    this.viewerOrigin_ = null;
```
Comment (dvoytenko) [inline src/service/viewer-impl.js]: Remove. The definition below is the correct one.

File: src/service/viewer-impl.js [hunk 18]
```diff
@@ -257,6 +267,22 @@ export class Viewer {
...
+    this.viewerOrigin_ = new Promise(resolve => {
+      if (!this.isEmbedded()) {
+        // Viewer is only determined for iframed documents at this time.
+        resolve('');
+      } else if (this.win.location.ancestorOrigins &&
+          this.win.location.ancestorOrigins.length > 0) {
+        resolve(this.win.location.ancestorOrigins[0]);
+      } else {
+        // Race to resolve with a timer.
+        timer.delay(() => { resolve(''); }, VIEWER_ORIGIN_TIMEOUT_);
```
Comment (dvoytenko) [inline src/service/viewer-impl.js]: Just this will suffice:
`timer.delay(() => resolve(''), VIEWER_ORIGIN_TIMEOUT_);`

File: src/service/viewer-impl.js [hunk 19]
```diff
@@ -495,6 +521,16 @@ export class Viewer {
   }
 
   /**
+   * Returns the promise that resolves to URL representing the origin of the
+   * viewer. If the document is not embedded or if a viewer origin can't be
+   * found, empty string is returned.
+   * @return {Promise<string>}
```
Comment (dvoytenko) [inline src/service/viewer-impl.js]: `!Promise`

File: test/functional/test-viewer.js [hunk 20]
```diff
@@ -450,4 +450,42 @@ describe('Viewer', () => {
...
+      const viewer = new Viewer(windowApi);
+      const result = viewer.getViewerOrigin().then(viewerOrigin => {
+        expect(viewerOrigin).to.equal('https://foobar.com');
+      });
+      viewer.setMessageDeliverer(() => {}, 'https://foobar.com');
+      return result;
+    });
+
+    it('should return empty string if handshake does not happen', () => {
+      windowApi.parent = {};
```
Comment (dvoytenko) [inline test/functional/test-viewer.js]: Please use fakeClock here - we don't want the test to actually wait for 1 second.

**Comment by dvoytenko** [comment]
Looking good - just a couple more comments.

File: test/functional/test-viewer.js (hunk 20 above)
Comment (avimehta) [inline test/functional/test-viewer.js]: done.

File: src/service/viewer-impl.js (hunk 19 above)
Comment (avimehta) [inline src/service/viewer-impl.js]: done.

File: src/service/viewer-impl.js (hunk 18 above)
Comment (avimehta) [inline src/service/viewer-impl.js]: done

File: src/service/viewer-impl.js (hunk 17 above)
Comment (avimehta) [inline src/service/viewer-impl.js]: done

**Comment by dvoytenko** [comment]
LGTM. Is this ready to be merged?

**Comment by avimehta** [comment]
Yes sir
